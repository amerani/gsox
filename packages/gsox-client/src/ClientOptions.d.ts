declare class ClientOptions {
      public host: string;
      public port: number;
      public routes: {
            graphql: string
            webhook: string,
      };
      public ws: any;
}
