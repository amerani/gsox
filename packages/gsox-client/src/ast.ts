import {
      DocumentNode,
      FieldNode,
      NameNode,
      OperationDefinitionNode,
      SelectionNode,
      SelectionSetNode,
    } from "graphql";

class SubscriptionDocument implements DocumentNode {
      public kind: "Document" = "Document";
      public definitions: OperationDefinitionNode[];
      constructor(definitions) {
            this.definitions = definitions;
            this.kind = "Document";
      }
}

class OperationNode implements OperationDefinitionNode {
      public kind: "OperationDefinition";
      public operation: "subscription";
      public selectionSet: SelectionSetNode;
      constructor(selectionSet) {
            this.selectionSet = selectionSet;
            this.kind = "OperationDefinition";
            this.operation = "subscription";
      }
}

class SelectionSet implements SelectionSetNode {
      public kind: "SelectionSet";
      public selections: FieldNode[];
      constructor(selections: FieldNode[]) {
            this.selections = selections;
            this.kind = "SelectionSet";
      }
}

class NameNodeImp implements NameNode {
      public kind: "Name";
      public value: string;
      constructor(value) {
            this.value = value;
            this.kind = "Name";
      }
}

class Field implements FieldNode {
      public kind: "Field";
      public name: NameNode;
      public selectionSet?: SelectionSetNode;
      constructor(name, selectionSet?) {
            this.name = new NameNodeImp(name);
            this.selectionSet = selectionSet;
            this.kind = "Field";
      }
}

export { Field, SelectionSet, OperationNode, SubscriptionDocument };
