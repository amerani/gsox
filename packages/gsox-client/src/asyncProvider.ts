import { InMemoryCache } from "apollo-cache-inmemory";
import { ApolloClient} from "apollo-client";
import { WebSocketLink } from "apollo-link-ws";
import { gql } from "apollo-server-express";
import { SubscriptionClient } from "subscriptions-transport-ws";
import * as ws from "ws";

const port = 3000;
const wsClient = new SubscriptionClient(`ws://localhost:${port}/graphql`, { reconnect: true }, ws);
const client = new ApolloClient({
      cache: new InMemoryCache({}),
      link: new WebSocketLink(wsClient),
});
client.subscribe({
      query: gql`
      subscription {
            msg {
                  type
                  id

            }
            }`,
}).subscribe(
      console.log,
      console.log,
      console.log,
);
