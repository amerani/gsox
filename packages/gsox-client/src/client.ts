import { InMemoryCache } from "apollo-cache-inmemory";
import ApolloClient from "apollo-client";
import { ApolloLink } from "apollo-link";
import { WebSocketLink } from "apollo-link-ws";
import { SubscriptionClient } from "subscriptions-transport-ws";
const GRAPHQL_ENDPOINT = "ws://localhost:3000/graphql";

const wsClient = new SubscriptionClient(GRAPHQL_ENDPOINT, {
  reconnect: true,
});

const wsLink = new WebSocketLink(wsClient);

export const client = new ApolloClient({
  cache: new InMemoryCache({}),
  connectToDevTools: true,
  link: ApolloLink.from([wsLink]),
  ssrForceFetchDelay: 100,
});
