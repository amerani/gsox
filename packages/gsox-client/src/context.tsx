import * as React from "react";

const StreamContext = React.createContext<{client:any}>({client:null});

export class StreamProvider extends React.Component<any> {
      render() {
            const { client, children } = this.props;
            return <StreamContext.Provider
                        value={{client}}
                        children={children}
                   />
      }
}

export class StreamConsumer extends React.Component<any> {
      private client = null;
      private subscription:ZenObservable.Subscription = null;
      state = {
            data: null,
            error: null,
            loading: true
      }

      componentWillUnmount() {
            this.subscription.unsubscribe();
            delete this.subscription;
      }

      init = (types) => this.subscription = this.subscription ||
            types.map(type =>
                  this.client.subscribe(type)
                  .subscribe({
                        next: ({data}) => this.setState({data, loading: false}),
                        error: (error) => this.setState({error, loading: false})
                  })
      )
      render() {
            const { children, types } = this.props as {children: (s)=>React.ReactNode, types:any};
            return <StreamContext.Consumer children={({client}):any => {
                  this.client = client;
                  this.init(types);
                  return children(this.state);
            }} />
      }
}