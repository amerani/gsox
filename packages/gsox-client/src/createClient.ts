import "reflect-metadata";
import gql from "graphql-tag";
import { InMemoryCache } from "apollo-cache-inmemory";
import ApolloClient from "apollo-client";
import { ApolloLink } from "apollo-link";
import { WebSocketLink } from "apollo-link-ws";
import { SubscriptionClient } from "subscriptions-transport-ws";
import { typeSub } from "@gsox/schema";

function createClient(options) {
      const { ws, host, port, routes } = options;
      const GRAPHQL_ENDPOINT = `ws://${host}:${port}${routes.graphql}`;
      const wsClient = new SubscriptionClient(GRAPHQL_ENDPOINT, { reconnect: true }, ws);

      const wsLink = new WebSocketLink(wsClient);

      const client = new ApolloClient({
        cache: new InMemoryCache(),
        connectToDevTools: true,
        link: ApolloLink.from([wsLink]),
        ssrForceFetchDelay: 100,
      });

      return {
            rawClient: client,
            subscribe: (T) => client.subscribe({
                  query: gql(typeSub(T))
            })
      }
}

export {
      createClient
}