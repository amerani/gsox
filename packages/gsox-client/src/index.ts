export * from "./createClient";
export * from "./context";
export * from "./typeGraph";
