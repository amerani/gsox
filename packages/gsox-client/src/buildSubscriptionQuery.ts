import { Field, OperationNode, SelectionSet, SubscriptionDocument } from "./ast";

function toGql(fields) {
    const gFields = [];
    for (let i = 0; i < fields.length; i++) {
        const field = new Field(fields[i].name);
        if (fields[i].children.length > 0) {
            field.selectionSet = new SelectionSet(toGql(fields[i].children));
        }
        gFields.push(field);
    }
    return gFields;
}

function build(typeName, fields) {
    const gqlFields = toGql(fields);
    const gqlSub = new Field(typeName, new SelectionSet(gqlFields));
    const op = new OperationNode(new SelectionSet([gqlSub]));
    const doc = new SubscriptionDocument([op]);
    return doc;
}

export { build };
