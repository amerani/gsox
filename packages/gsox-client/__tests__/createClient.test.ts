import { createClient } from "../src/createClient";

test("should create client", () => {
      expect(createClient({ } as any)).toMatchSnapshot();
});

test("should create client with options", () => {
      const options:any = { host: "localhost", port: 3000, routes: { graphql: "/gql" }, inject: { } };
      expect(createClient(options)).toMatchSnapshot();
});
