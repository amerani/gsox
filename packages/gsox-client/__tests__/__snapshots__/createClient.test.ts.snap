// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create client 1`] = `
Object {
  "rawClient": ApolloClient {
    "cache": InMemoryCache {
      "addTypename": true,
      "config": Object {
        "addTypename": true,
        "dataIdFromObject": [Function],
        "fragmentMatcher": HeuristicFragmentMatcher {},
        "storeFactory": [Function],
      },
      "data": ObjectCache {
        "data": Object {},
      },
      "optimistic": Array [],
      "silenceBroadcast": false,
      "typenameDocumentCache": WeakMap {},
      "watches": Array [],
    },
    "defaultOptions": Object {},
    "disableNetworkFetches": true,
    "link": ApolloLink {
      "request": [Function],
    },
    "mutate": [Function],
    "query": [Function],
    "queryDeduplication": true,
    "reFetchObservableQueries": [Function],
    "resetStore": [Function],
    "resetStoreCallbacks": Array [],
    "ssrMode": false,
    "store": DataStore {
      "cache": InMemoryCache {
        "addTypename": true,
        "config": Object {
          "addTypename": true,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "storeFactory": [Function],
        },
        "data": ObjectCache {
          "data": Object {},
        },
        "optimistic": Array [],
        "silenceBroadcast": false,
        "typenameDocumentCache": WeakMap {},
        "watches": Array [],
      },
    },
    "version": "2.3.7",
    "watchQuery": [Function],
  },
  "subscribe": [Function],
}
`;

exports[`should create client with options 1`] = `
Object {
  "rawClient": ApolloClient {
    "cache": InMemoryCache {
      "addTypename": true,
      "config": Object {
        "addTypename": true,
        "dataIdFromObject": [Function],
        "fragmentMatcher": HeuristicFragmentMatcher {},
        "storeFactory": [Function],
      },
      "data": ObjectCache {
        "data": Object {},
      },
      "optimistic": Array [],
      "silenceBroadcast": false,
      "typenameDocumentCache": WeakMap {},
      "watches": Array [],
    },
    "defaultOptions": Object {},
    "disableNetworkFetches": true,
    "link": ApolloLink {
      "request": [Function],
    },
    "mutate": [Function],
    "query": [Function],
    "queryDeduplication": true,
    "reFetchObservableQueries": [Function],
    "resetStore": [Function],
    "resetStoreCallbacks": Array [],
    "ssrMode": false,
    "store": DataStore {
      "cache": InMemoryCache {
        "addTypename": true,
        "config": Object {
          "addTypename": true,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "storeFactory": [Function],
        },
        "data": ObjectCache {
          "data": Object {},
        },
        "optimistic": Array [],
        "silenceBroadcast": false,
        "typenameDocumentCache": WeakMap {},
        "watches": Array [],
      },
    },
    "version": "2.3.7",
    "watchQuery": [Function],
  },
  "subscribe": [Function],
}
`;
